### DISCUSSION TITLE
Existing ERT test with Eglot failing with recent update from `project-name` not found

### DISCUSSION THREAD

**Discussion by andyleejordan** [joaotavora/eglot#1188] [Q&A]
I'm not sure what's going on here. I have a CI test for the PowerShell LSP server that asserts it works with Eglot, [here's the ERT script](https://github.com/PowerShell/PowerShellEditorServices/blob/d4b642586768391eb0e9889ff56f501cea09cc48/test/emacs-test.el). It suddenly started failing! Here's the snippet of the CI failure:

```
 Contacting host: elpa.gnu.org:443
 Parsing tar file... 
 Parsing tar file...done
 Extracting... \ 
 Extracting...done
   INFO     Scraping files for project-autoloads.el... 
   INFO     Scraping files for project-autoloads.el...done
 Checking /home/runner/.emacs.d/elpa/project-0.9.8...
 Compiling /home/runner/.emacs.d/elpa/project-0.9.8/project-autoloads.el...
 Compiling /home/runner/.emacs.d/elpa/project-0.9.8/project-pkg.el...
 Compiling /home/runner/.emacs.d/elpa/project-0.9.8/project.el...
 
 In toplevel form:
 project.el:255:9: Warning: ‘project-roots’ is an obsolete function (as of
     0.3.0); use ‘project-root’ instead.
 project.el:257:16: Warning: ‘project-roots’ is an obsolete generic function
     (as of 0.3.0); use ‘project-root’ instead.
 
 In project-find-file:
 project.el:998:6: Warning: project-find-file-in called with 4 arguments, but
     accepts only 3
 
 In project-or-external-find-file:
 project.el:1018:6: Warning: project-find-file-in called with 4 arguments, but
     accepts only 3
 
 In project-query-replace-regexp:
 project.el:1232:4: Warning: Unused lexical variable
     `query-replace-read-from-regexp-default'
 Done (Total of 1 file compiled, 2 skipped)
 Contacting host: elpa.gnu.org:443
 Parsing tar file... 
 Parsing tar file...done
 Extracting... \ 
 Extracting...done
   INFO     Scraping files for jsonrpc-autoloads.el... 
   INFO     Scraping files for jsonrpc-autoloads.el...done
 Checking /home/runner/.emacs.d/elpa/jsonrpc-1.0.17...
 Compiling /home/runner/.emacs.d/elpa/jsonrpc-1.0.17/jsonrpc-autoloads.el...
 Compiling /home/runner/.emacs.d/elpa/jsonrpc-1.0.17/jsonrpc-pkg.el...
 Compiling /home/runner/.emacs.d/elpa/jsonrpc-1.0.17/jsonrpc.el...
 Done (Total of 1 file compiled, 2 skipped)
 Parsing tar file... 
 Parsing tar file...done
 Extracting... \ 
 Extracting...done
   INFO     Scraping files for eglot-autoloads.el... 
   INFO     Scraping files for eglot-autoloads.el...done
 Checking /home/runner/.emacs.d/elpa/eglot-1.13...
 Compiling /home/runner/.emacs.d/elpa/eglot-1.13/eglot-autoloads.el...
 Compiling /home/runner/.emacs.d/elpa/eglot-1.13/eglot-pkg.el...
 Compiling /home/runner/.emacs.d/elpa/eglot-1.13/eglot.el...
 
 In end of data:
 eglot.el:1223:21: Warning: the function ‘project-name’ is not known to be
     defined.
 Done (Total of 1 file compiled, 2 skipped)
 Package ‘eglot’ installed.
 Running 1 tests (2023-03-22 22:47:11+0000, selector ‘t’)
 Test powershell-editor-services backtrace:
   project-name((vc . "~/work/PowerShellEditorServices/PowerShellEditor
   eglot--connect((powershell-mode) (vc . "~/work/PowerShellEditorServi
   apply(eglot--connect ((powershell-mode) (vc . "~/work/PowerShellEdit
   apply(apply (eglot--connect ((powershell-mode) (vc . "~/work/PowerSh
   (setq value-133 (apply fn-131 args-132))
   (unwind-protect (setq value-133 (apply fn-131 args-132)) (setq form-
   (if (unwind-protect (setq value-133 (apply fn-131 args-132)) (setq f
   (let (form-description-135) (if (unwind-protect (setq value-133 (app
   (let ((value-133 'ert-form-evaluation-aborted-134)) (let (form-descr
   (let* ((fn-131 #'apply) (args-132 (condition-case err (let ((signal-
   (save-current-buffer (set-buffer (find-file-noselect test-script)) (
   (let* ((repo (project-root (project-current))) (start-script (expand
   (let ((lexical-binding t)) (let* ((repo (project-root (project-curre
   (closure (t) nil (let ((lexical-binding t)) (let* ((repo (project-ro
   ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
   ert-run-test(#s(ert-test :name powershell-editor-services :documenta
   ert-run-or-rerun-test(#s(ert--stats :selector t :tests [...] :test-m
   ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
   ert-run-tests-batch(nil)
   ert-run-tests-batch-and-exit()
   command-line-1(("-l" "ert" "-l" "test/emacs-test.el" "-f" "ert-run-t
   command-line()
   normal-top-level()
 Test powershell-editor-services condition:
     (void-function project-name)
    FAILED  1/1  powershell-editor-services (0.068108 sec)
 
 Ran 1 tests, 0 results as expected, 1 unexpected (2023-03-22 22:47:11+0000, 0.250636 sec)
```

That `(void-function project-name)` is weird! I looked through Eglot 1.13 and Project 0.9.8 and...well it exists. I first tried adding `(require 'project)` to the test script. Didn't work. Finally I just copied its definition from `project.el`:

```
(cl-defgeneric project-name (project)
  (file-name-base (directory-file-name (project-root project))))
```

into the test script...and it's working again. But that shouldn't be necessary! @joaotavora it's been forever and a half since I got to hangout in Emacs land, do you have any idea what's up?

**Comment by joaotavora** [comment]
Hi. I'm not 100% of this, but your usage pattern points to some familiar cause witht htese errors.  The emacs version you're using has as built-in `lisp/progmodes/project.el`.  That version is old and doesn't have `project-name`.  When starting your test suite in your CI system, you ask Emacs to install Eglot, the ELPA package, which depends on a newer version of `project.el` which will also be installed and compiled.
The fact that it's being compiled is visible in these lines:
```
hecking /home/runner/.emacs.d/elpa/project-0.9.8...
 Compiling /home/runner/.emacs.d/elpa/project-0.9.8/project.el...
```

But in Lisp, "compiling" doesn't equal "loading", so that by itself won't load in the new project.

For some reason, your CI run loaded caused the "old" `project.el` to be loaded, i.e. that 'load' or 'require' somehow happened before the package system set itself up, and thus the `require 'project` in Eglot.el had no effect.  I can't be of more help without looking into your CI scripts and how they fire up Emacs.

BTW This problem doesn't normally happen in interactive usage of Emacs, because `package-initialize` is run pretty early on. But it's not extremely hard to trigger either, as you've just discovered.

**Reply by andyleejordan** [reply]
Thanks for the help! I'm using [purcell/setup-emacs@master](https://github.com/purcell/setup-emacs) to install and test with Emacs 28.2 via Nix...which I thought would come with project.el 0.9.8 but you must be right about some older project.el being loaded first, so  I'll sleuth around some more.

**Reply by joaotavora** [reply]
Ah. This Nix package manager has been known to mess up Eglot installations.  Hope you can find out the casue.  There are some related issues around here for the same `project-name` thing (though I can't seem to find them right now).

**Reply by andyleejordan** [reply]
Grr...I was hopeful it was a matter of forgetting to `(setq load-prefer-newer t)` but no. Then I also switched to a simple `use-package` with `:ensure` setup to be like "yo install the updated `project.el` and load it please and thank you" but no. So I checked your CI to see what you're using to install Emacs in the GitHub workflow...also purcell/setup-emacs. 😳

... I think I have to code in a forced "update installed packages" thing because `:ensure t` is just going "hey you got it!"

**Reply by joaotavora** [reply]
`use-package` might be doing much confusing magic.  I would avoid it, unless I'm absolutely sure what that macro does.  

You should be able to get by with a call to `emacs -Q --batch -f package-refresh-contents --eval '(package-install (quote eglot))'` followed by a call to `emacs -Q --batch -f package-initialize -l your-test-file.el -f ert-run-tests-batch-and-exit` where you run the tests.

**Reply by andyleejordan** [reply]
OMG thank you. I think the big trick was `-Q` (which implies `--no-site-lisp` and I was missing that, turns out `--batch` only implies `-q` which is `--no-init-file` and completely different). And I'm taking your advice and installing `eglot` in that manner and _then_ running the ERT script (which still goes and adds MELPA and installs `powershell` so the rest of the stuff works). THANK YOU.

**Reply by joaotavora** [reply]
> I think the big trick was -Q (which implies --no-site-lisp and I was missing that, turns out --batch only implies -q which is --no-init-file and completely different). 

They're not _that_ different.  `-Q` implies `-q`, but not the other around.  The difference is only relevant if you have a "site lisp" file, which are relatively rare nowadays.  Such files are used for site-wide configuration of Emacs (think multiple users using the same machine).  It's not very relevant in the CI.

I think your problem was that:

1.  You were installing Eglot and dependencies from scratch and running the tests in the same session.
2. The above  only works if you can ensure that `project.el` is _not_ required before installing a new version.

My conjecture is that you violated 2 somewhere somehow.  Two Emacs sessions makes sure you're safe from that pitfall.

**Reply by andyleejordan** [reply]
Ahh yes, yes I see. And I couldn't figure out a good way to ensure `project.el` was installed and _updated_ in that session, but the two Emacs sessions works great! (I tried the `auto-package-update` package...I tried running `package-reinstall` but it always complained `project` wasn't installed...which was kinda the crux of the issue because it was "bundled" not installed). Anyway, thank you a million for your help! I always appreciate it.
